### PR TITLE
[docs/train] Add Result guide

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

This adds a guide on how to use the Ray Train `Result` object.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/train/user-guides/results.rst
Comment (matthewdeng) [inline doc/source/train/user-guides/results.rst:1]: This is awesome, one high level suggestion for extending this is to add some more color to what the user should do with these attributes.

As a user, what can/should I do with the metrics, checkpoints, etc.? We can guide them towards common steps, such as visualizing metrics with TensorBoard, or using the checkpoint for prediction.

File: doc/source/train/user-guides/results.rst
Comment (matthewdeng) [inline doc/source/train/user-guides/results.rst:13]: How about grouping "Viewing metrics" as one header and having sub-headers for latest/all metrics?

That way we can add some general context on what to do with metrics at the top level, that is applicable to all.

File: doc/source/train/user-guides/results.rst
Comment (matthewdeng) [inline doc/source/train/user-guides/results.rst:38]: Same as above but for checkpoints, there is a difference in ways to retrieve checkpoints, but should have a common section for what the user might do with the checkpoints.

File: doc/source/train/user-guides/results.rst
Comment (matthewdeng) [inline doc/source/train/user-guides/results.rst:16]: These don't seem to be rendering properly.

File: doc/source/train/user-guides/results.rst
Comment (matthewdeng) [inline doc/source/train/user-guides/results.rst:69]: What does "inspect" mean here?

File: doc/source/train/user-guides/results.rst
Comment (krfricke) [reply doc/source/train/user-guides/results.rst:16]: Yeah I've tried a few things and couldn't find the right keyword so far

Comment (krfricke) [reply doc/source/train/user-guides/results.rst]: Found it - for the attributes `ray.air.Result.metrics` etc has to be used (not `result.Result`...)

File: doc/source/train/user-guides/results.rst
Comment (krfricke) [reply doc/source/train/user-guides/results.rst:1]: Added a bit - keeping it concise for now to not clutter the page, but happy to add more references.

File: doc/source/train/user-guides/results.rst
Comment (matthewdeng) [inline doc/source/train/user-guides/results.rst:5]: Can you add a section that explains way to create a Result? The part that's missing right now is that you can also load a previously created result from `Result.from_path` (this might go hand-in-hand with the storage location section).

File: doc/source/train/user-guides/results.rst
Comment (krfricke) [reply doc/source/train/user-guides/results.rst:95]: Updated.

For result restoration, let's wait until the refactor is landed.

File: doc/source/train/user-guides/results.rst
Comment (matthewdeng) [inline doc/source/train/user-guides/results.rst:39]:
```suggestion
Use :attr:`Result.metrics_dataframe <ray.train.Result.metrics_dataframe>` to retrieve
```

File: doc/source/train/key-concepts.rst
Comment (matthewdeng) [inline doc/source/train/key-concepts.rst:76]:
```suggestion
Calling ``Trainer.fit()`` returns a :class:`Result <ray.train.Result>` object, which includes
```

**Review by matthewdeng — APPROVED** [review]

**Review by c21 — APPROVED** [review]

**Pull request merged**
